pkg/disruptors/controller.go

```diff
@@ -96,42 +105,76 @@ func (c *agentController) InjectDisruptorAgent(ctx context.Context) error {
 // ExecCommand executes a command in the targets of the AgentController and reports any error
 func (c *agentController) ExecCommand(ctx context.Context, cmd []string) error {
 	// visit each target with the same command
-	return c.Visit(ctx, func(corev1.Pod) ([]string, error) {
-		return cmd, nil
+	return c.Visit(ctx, func(corev1.Pod) (VisitCommands, error) {
+		return VisitCommands{Exec: cmd}, nil
 	})
 }
 
 // Visit allows executing a different command on each target returned by a visiting function
-func (c *agentController) Visit(_ context.Context, visitor func(corev1.Pod) ([]string, error)) error {
-	var wg sync.WaitGroup
-	// ensure errors channel has enough space to avoid blocking gorutines
-	errors := make(chan error, len(c.targets))
-	for _, pod := range c.targets {
-		wg.Add(1)
-		// attach each container asynchronously
-		go func(pod corev1.Pod) {
-			// get the command to execute in the target
-			cmd, err := visitor(pod)
-			if err != nil {
-				errors <- fmt.Errorf("error building command for pod %s: %w", pod.Name, err)
-			}
+func (c *agentController) Visit(ctx context.Context, visitor func(corev1.Pod) (VisitCommands, error)) error {
+	// if there are no targets, nothing to do
+	if len(c.targets) == 0 {
+		return nil
+	}
 
-			_, stderr, err := c.helper.Exec(pod.Name, "xk6-agent", cmd, []byte{})
-			if err != nil {
-				errors <- fmt.Errorf("error invoking agent: %w \n%s", err, string(stderr))
-			}
+	execContext, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
-			wg.Done()
-		}(pod)
+	// ensure errCh channel has enough space to avoid blocking gorutines
+	errCh := make(chan error, len(c.targets))
+	for _, pod := range c.targets {
+		pod := pod
+		// visit each target asynchronously
+		go func() {
+			errCh <- func(pod corev1.Pod) error {
+				// get the command to execute in the target
+				visitCommands, err := visitor(pod)
+				if err != nil {
+					return fmt.Errorf("unable to get command for pod %q: %w", pod.Name, err)
+				}
+
+				_, stderr, err := c.helper.Exec(execContext, pod.Name, "xk6-agent", visitCommands.Exec, []byte{})
+
+				// if command failed, ensure the agent execution is terminated
+				if err != nil && visitCommands.Cleanup != nil {
+					// we ignore errors because k6 was cancelled, so there's no point in reporting
+					// use a fresh context because the exec context may have been cancelled or expired
+					//nolint:contextcheck
+					_, _, _ = c.helper.Exec(context.TODO(), pod.Name, "xk6-agent", visitCommands.Cleanup, []byte{})
+				}
+
+				// if the context is cancelled, it is reported in the main loop
+				if err != nil && !errors.Is(err, context.Canceled) {
```

Comment:
@roobre I added this check to prevent the exec context cancellation to be reported.